README
NAME
    CGI::Application::Search - Base class for CGI::App Swish-e site engines

SYNOPSIS
        package My::Search;
        use base 'CGI::Application::Search';
        
        sub cgiapp_init {
            my $self = shift;
            $self->param(
                'SWISHE_INDEX' => 'my-swishe.index',
                'TEMPLATE'     => 'search_results.tmpl',
            );
        }

        #let the user turn context highlighting off
        sub cgiapp_prerun {
            my $self = shift;
            $self->param('HIGHLIGHT' => 0)
                if($self->query->param('highlight_off'));
        }

        1;

DESCRIPTION
    A CGI::Application based control module that uses Swish-e API in perl
    (<http://swish-e.org>) to to perform searches on a swish-e index of
    documents.

  Features
    * Sub-Classable. Unlike the Perl examples that come with swish-e, this
    is not a script, and can be customized without modifiying the original
    so that several sites may share the same underlying code.
    * Uses CGI::Application::Plugin::AnyTemplate to allow flexibility in
    template engine choice (HTML::Template, Template-Toolkit or Petal).
    * Built-in templates to use out of box or as examples for your own
    templates
    * HiLighted search results
    * HiLighted pages linked from search results
    * AJAX results sent to page without need of a page reload
    * AJAX powered 'auto-suggest' to give the user list of choices available
    for search

TUTORIAL
    If this is your first time using Swish-e (or you think you need a
    refresher) or if you want step-by-step instructions on how to use the
    AJAX capabilities of this module, then please see the "Tutorial".

RUN_MODES
    The start_mode is show_search.

  show_search()
    This method will load the template pointed to by the TEMPLATE param
    (falling back to a default internal template if none is specified) and
    display it to the user. It will 'associate' this template with $self so
    that any parameters in $self->param() are also accessible to the
    template. It will also use HTML::FillInForm to fill in the search form
    with the previously selected parameters (unless it's a
    'non-page-refresh' AJAX search).

  perform_search()
    This is where the meat of the searching is performed. We create a
    SWISH::API object on the SWISHE_INDEX and create the query for the
    search based on the value of the *keywords* parameter in CGI and any
    other EXTRA_PARAMETERS. The search is executed and if HIGHLIGHT is true
    we will use HTML::HiLiter to highlight it and then format the results,
    only showing PER_PAGE number of elements A paging list is also shown for
    navigating through the results. Then we will return to the show_search()
    method for displaying everything.

  highlight_remote_page
    This run mode will fetch a remote page (with either a relative, or
    absolute URL using the "url" Query param) and highlight the keywords
    used in the search on that page (the "keywords" Query param) using the
    HIGHLIGHT_TAG, HIGHLIGHT_CLASS or HIGHLIGHT_COLORS options. This run
    mode is best used in the links of the search results listing.

        <a href="?rm=highlight_remote_page;url=http%3A%2F%2Fexample.com%2Fabout_us%2Findex.html;keywords=Us">about us</a>

  highlight_local_page
    This run mode will fetch a local page (only allowing relative files
    based in the DOCUMENT_ROOT config var and the path using the "path"
    Query param) and highlight the keywords used in the search on that page
    (the "keywords" Query param) using the HIGHLIGHT_TAG, HIGHLIGHT_CLASS or
    HIGHLIGHT_COLORS options. This run mode is best used in the links of the
    search results listing.

        <a href="?rm=highlight_local_page;path=%2Fabout_us%2Findex.html;keywords=Us">about us</a>

  suggestions
    This run mode will return an AJAX listing of words that should be
    suggested to the user for the words that they have typed so far. It uses
    the "suggested_words()" method to actually choose which words to send
    back.

OTHER METHODS
    Most of the time you will not need to call the methods that are
    implemented in this module. But in some cases customizing the templates
    is not enought. If so, it might be prudent to override or extend these
    methods in your derived class.

  new()
    We simply override and extend the CGI::Application new() to setup our
    defaults.

  setup()
    Here's were we setup our run modes. If you override this method, make
    sure you also call it in your base class

        sub setup {
            my $self = shift;
            # do your thing
            ...
            $self->SUPER::setup();
        }

  generate_search_query($keywords)
    This method is used to generate the query for swish-e from the $keywords
    (by default the 'keywords' CGI parameter), as well as any
    EXTRA_PROPERTIES that are present.

    If you wish to generate your own search query then you should override
    this method. This is common if you need to have access/authorization
    control that will need to be taken into account for your search. (eg,
    anything under /protected can't be seen by someone not logged in).

    Please see the swish-e documentation on the exact syntax for the query.

  suggested_words($word)
    This object method is used by the AUTO_SUGGEST flag to return the words
    that should be suggested to the user after they have typed a $word. It
    returns an array reference of those words.

    By default it will just look for words in the AUTO_SUGGEST_FILE that
    begin with $word. If you need more performance or flexibility (eg,
    storing your words in a database and querying for them) you are
    encouraged to override this method.

CONFIGURATION
    There are several configuration parameters that you can set at any time
    (using "param()" in your cgiapp_init, or PARAMS hash in new()) before
    the run mode is called that will affect the search and display of the
    results. They are:

  SWISHE_INDEX
    This is the swishe index used for the searches. The default is
    'data/swish-e.index'. You will probably set this every time.

  AJAX
    This is a boolean indicating whether or not a non-page-refresh AJAX
    search will be permitted.

    Please see the "Tutorial" for more information on how to use the AJAX
    capabilities of this module.

  TEMPLATE
    The name of the search interface template. Default templates are
    included with this distribution and will be used if you don't specify
    one. Which default template is used depends on which TEMPLATE_TYPE you
    are using (*HTMLTemplate* or *TemplateToolkit*) and whether or not the
    AJAX flag is true.

    These sample templates are installed with the module, or you can view
    them by looking under the templates/ directory of the source
    distribution (*.tar.gz*).

    Please see "TEMPLATE USAGE" for more information on which variables are
    passed into your template.

  TEMPLATE_TYPE
    This module uses CGI::Application::Plugin::AnyTemplate to allow
    flexibility in choosing which templating system to use for your search.
    This works especially well when you are trying to integrate the Search
    into an existing app with an existing templating structure.

    This value is passed to the "$self->template->config()" method as the
    "default_type". By default it is 'HTMLTemplate'. Please see
    CGI::Application::Plugin::AnyTemplate for more options.

    If you want more control of configuration for the template the it would
    probably best be done by subclassing CGI::Application::Search and
    passing your desired params to "$self->template->config".

  PER_PAGE
    How many search result items to display per page. The default is 10.

  HIGHLIGHT
    Boolean indicating whether or not we should highlight the description
    given to the templates. The default is true.

  HIGHLIGHT_TAG
    The HTML tag used to surround the highlighted context. The default is
    "strong".

  HIGHLIGHT_CLASS
    The class attribute of the HIGHLIGHT_TAG HTML tag. This is useful when
    you want to dictacte the style through a CSS style sheet. If given, this
    value will override that of HIGHLIGHT_COLORS. It has no value by
    default.

  HIGHLIGHT_COLORS
    This is an array ref of acceptable HTML colors. If provided, it will
    highlight each matching word/phrase in an alternating style. For
    instance, if given 2 colors, every other highlighted phrase would be a
    different color. By default it is an empty array.

  EXTRA_PROPERTIES
    This is an array ref of extra properties used in the search. By default,
    the module will only use the value of the 'keywords' parameter coming in
    the CGI query. If anything is provided as an extra property then it will
    be added to the query used in the search.

    An example: You have some of you pages designated into categories. You
    want the user to have the option of narrowing his results by category.
    You add the word 'category' to the 'EXTRA_PROPERTIES' list and then you
    add a 'category' form element that the user has the option of giving a
    value to your search form. If the user gives that element a value, then
    it will be seen and applied to the search. This will also only work if
    you have the 'category' element defined for your documents (see *SWISH-E
    Configuration* and 'MetaNames' in the swish-e.org SWISH-CONF
    documentation).

    By default, this list is empty.

  EXTRA_RANGE_PROPERTIES
    This is almost exactly like the "EXTRA_PROPERTIES" above except that
    instead of searching for the given properties as simple strings, we will
    use a range. Since ranges need to values, if you're searching for the
    "foo" property, then you need to have a "foo_start" and a "foo_end"
    value coming from the query. So if "foo" is in your
    "EXTRA_RANGE_PROPERTIES" and you have a CGI query string like this:

        ?foo_start=123&foo_end=234

    Then we will a generate a Swish-E query that looks something like this:

      -L foo 123 234

  DESCRIPTION_LENGTH
    This is the maximum length for the context (in chars) that is displayed
    for each search result. The default is 250 characters.

  DESCRIPTION_CONTEXT
    This is the number of words on either side of the searched for words and
    phrases (keywords) that will be displayed as part of the description. If
    this is 0, then the entire description will be displayed. The default is
    0.

    NOTE: This directive will cause Search to use Text::Context, which can
    be slow and CPU intensive at times. These computations may prove to be
    too much for some servers (eg, a shared hosting environment).

  AUTO_SUGGEST
    If true, then this will allow the broswer to give suggestions to the
    user as they type. To use this, you must either use the
    AUTO_SUGGEST_FILE configuration option, or override the
    "suggested_words()" method.

    You template must also have the appropriate JavaScript code. Please see
    the "Tutorial" for more details.

  AUTO_SUGGEST_FILE
    The name of the file where the suggested words are stored. These words
    should be in alphabetical order with one word per line.

  AUTO_SUGGEST_CACHE
    A boolean indicating whether or not the results of the AUTO_SUGGEST_FILE
    should be cached in memory or not. This will save repeated file accesses
    when used in a persistant environment.

  AUTO_SUGGEST_LIMIT
    An integer count of the most suggestions to show the user at a time.
    This is useful when you don't want to overwhelm the end user and take
    over their screen with all of your helpful suggestions.

  DOCUMENT_ROOT
    This is the root directory to use when looking for files when using the
    "highlight_local_page" run mode.

TEMPLATE USAGE
    Sample templates are included with this distribution. These sample
    templates are installed with the module, or you can view them by looking
    under the templates/ directory of the source distribution (*.tar.gz*).

    Please feel free to copy and change them in what ever way you see fit.
    To help give you more information to display (or not display, depending
    on your preference) the following variables are available for your
    templates:

  Global Tmpl Vars
    These variables are available throughout the templates and contain
    information related to the search as a whole:

    * ajax  A boolean indicating whether or not this search is an AJAX
            search or not. You can use this flag to exclude everything but
            your search results in your template.

    * url   The URL of this application. This is useful if you want to use
            the same templates in multiple applications, especially if you
            are using the AJAX capabilities since they require the URL to
            submit to.

    * searched
            A boolean indicating whether or not a search was performed.

    * keywords
            The exact string that was recieved by the server from the input
            named 'keywords'

    * elapsed_time
            A string representing the number of seconds that the search
            took. This will be a floating point number with a precision of
            3.

    * hits  This is an array of hashs (TMPL_LOOP in H::T) that contains one
            entry for each result returned (for the current page). Each
            entry contains the following keys:

            reccount
                    The "swishreccount" property of the results as indexed
                    by SWISH-E

            rank    The rank to the result as given by SWISH-E (the
                    "swishrank" property)

            title   The "swishtitle" property of the results as indexed by
                    SWISH-E

            path    The "swishdocpath" property of the results as indexed by
                    SWISH-E

            last_modified
                    The "swishlastmodified" property of the results as
                    indexed by SWISH-E and then formatted using
                    Time::Piece's "strftime()" method with a format string
                    of "%B %d, %Y".

            size    The "swishdocsize" property of the results as indexed by
                    SWISH-E and then formatted with Number::Format's
                    "format_bytes()" method.

            description
                    The "swishdescription" property of the results as
                    indexed by SWISH-E. If HIGHLIGHT is true, then this
                    description will also have search terms highlighted and
                    will only be, at most, DESCRIPTION_LENGTH characters
                    long.

    * pages This is an array of hashes (TMPL_LOOP in H::T) that contains
            paging information for the results. It contains the following
            keys:

            current A boolean indicating whether or not this iteration is
                    the current page or not.

            page_num
                    The integer number of the page.

    * first_page
            This is a boolean indicating whether or not this page of the
            results is the first or not.

    * last_page
            This is a boolean indicating whether or not this page of the
            results is the last or not.

    * prev_page
            The integer number of the previous page. Will be 0 if there is
            no previous page.

    * next_page
            The integer number of the next page. Will be 0 if there is no
            next page.

    * start_num
            This is the number of the first result on the current page

    * stop_num
            This is the number of the last result on the current page

    * total_entries
            The total number of results in their search, not the total
            number shown on the page.

OTHER NOTES
    *   If at any time prior to the execution of the 'perform_search' run
        mode you set the "$self->param('results')" parameter, a search will
        not be performed. Instead those results are returned. This is
        helpful when you decide in the "cgiapp_init" stage that this user
        does not have permissions to perform the desired search.

    *   You must use the *StoreDescription* setting in your Swish-e
        configuration file. If you don't you'll get an error when
        C::A::Search tries to retrieve a description for each hit.

AUTHOR
    Michael Peters <mpeters@plusthree.com>

    Thanks to Plus Three, LP (http://www.plusthree.com) for sponsoring my
    work on this module.

CONTRIBUTORS
    Sam Tregar <sam@tregar.com>
    Mark Stosberg <mark@summersault.com>



Changes
Revision history for Perl extension CGI::Application::Dispatch.

1.04 (Nov 06, 2007)
    - Fixed "associate" regression where data in $self->param wasn't
      getting sent to the templates.
    - Fixed broken Makefile.PL
    - Adding new EXTRA_RANGE_PROPERTIES to the available fields returned in the results

1.03 (Nov 05, 2007)
    - Changed generate_search_query to allow searching when no
      key words are specified. The search can be on other properties
      or for everything listed.

1.02 (Nov 05, 2007)
    - Now using swish-e 2.4.5
    - Added EXTRA_RANGE_PROPERTIES

1.00 (Jul 08 2006)
    - Added Test::LongString and Test::Class to build_requires
    - Fixed bug with URL generation not including PATH_INFO
    - Fix bug in Ajax TT template that didn't work in non-ajax browsers

0.10 (Nov 21, 2005)
    - Changes to Makefile.PL to make CPAN installation happier
    - documentation cleanup and clarifications

0.09 (Nov 17, 2005)
    - Default templates now fail more gracefully when the browser doesn't support AJAX
    - no longer need both AJAX and AUTO_SUGGEST to use just the AUTO_SUGGEST functionality
    - Makefile.PL not installs sample templates like Build.PL (this should make CPAN happy)

0.08 (Nov 16, 2005)
    - No code changes, just refactored out the repetitive tests using Test::Class
    - move tests and coverage

0.07 (Nov 13, 2005)
    - No code changes just fixed failing tests when installed via CPAN

0.06 (Nov 12, 2005)
    - added AJAX flag to turn on non-refresh searches
    - added AUTO_SUGGEST, AUTO_SUGGEST_FILE, AUTO_SUGGEST_CACHE and AUTO_SUGGEST_LIMIT
        functionality
    - minor doc fixes
    - fixed mispelling of 'global_var' => 'global_vars' in HTML::Template
        options
    - removed use of callbacks since they didn't really add anything
        and required newer versions of C::A
    - better and more complete sample templates that are installed with
        the application and useable when no others are provided.
    - split tutorial out into a separate file and include some setup
        info for AJAX usage.

0.05 (Jul 10, 2005)
    - renamed 'HIGHLIGHT_CONTEXT' to just 'HIGHLIGHT'
    - removed 'HIGHLIGT_STOP' and 'HIGHLIGHT_START' in favor of
        'HIGHLIGHT_TAG' and 'HIGHLIGHT_CLASS' and 'HIGHLIGHT_COLORS'
    - renamed 'CONTEXT_LENGTH' to 'DESCRIPTION_LENGTH'
    - default highlighting is now done with HTML::HiLiter (much faster)
    - DESCRIPTION_CONTEXT added so that highlighting is now independent of
        finding a matching 'context' inside the description

0.04 (Jul 08, 2005)
    - removed 'hit_' prefix from all vars in in the 'hits' loop for templates
    - almost 100% test coverage (99.6%)
    - minor bug fixes with built-in template
    - use Carp for all errors and warnings

0.03 (Jul 05, 2005)
    - changed from straight H::T to CGI::Application::Plugin::AnyTemplate
        for flexibility
    - better template documentation and minor template fixes
    - more complete test suite, still not 100% coverage, but closer

0.02 (May 23, 2005)
    - find and fix document mistakes - Sam Tregar <sam@tregar.com>
    - added tutorial for getting started with swishe - Sam Tregar
    - added TEMPLATE option to setup() - Sam Tregar

0.01 (May 18, 2005)
    - First released version


t/lib/MySearch.pm
package MySearch;
use base 'CGI::Application::Search';

sub cgiapp_init {
    my $self = shift;
    my $query = $self->query;

    # allow the extra property 'extra'
    $self->param(
        EXTRA_PROPERTIES => [qw(extra)],
    );

    # let the user turn context highlighting on
    # defaulting to off
    $self->param('HIGHLIGHT' => ( $query->param('hl') || 0 ));

    # let the user set the per page
    $self->param('PER_PAGE' => $query->param('per_page'));

    # let the user pick an index (not really a good idea in production)
    my $index = $query->param('index') || 't/conf/swish-e.index';
    $self->param(SWISHE_INDEX => $index);

    # let the user turn on context
    $self->param(DESCRIPTION_CONTEXT => ($query->param('context') || 0));
}

1;


t/lib/MySearch/HT.pm
package MySearch::HT;
use base 'MySearch';

sub cgiapp_prerun {
    my $self = shift;
    $self->param(
        TEMPLATE        => 'templates/search_results.tmpl',
        TEMPLATE_TYPE   => 'HTMLTemplate'
    );
}


1;


t/lib/MySearch/TT.pm
package MySearch::TT;
use base 'MySearch';

sub cgiapp_init {
    my $self = shift;
    $self->SUPER::cgiapp_init(@_);
    $self->param(
        TEMPLATE            => 'templates/search_results.tt',
        TEMPLATE_TYPE       => 'TemplateToolkit',
        TEMPLATE_CONFIG     => {
            DEBUG   => 1,
        },
    );
    $self->tmpl_path('.');
}


1;


Build.PL
use Module::Build;
use lib './t/lib';
Module::Build->new(
    module_name => 'CGI::Application::Search',
    license => 'perl',
    requires => {
        'perl'                                    => '5.6.1',
        'CGI::Application'                        => 3.2,
        'CGI::Application::Plugin::AnyTemplate'   => 0,
        'SWISH::API'                              => 0.04, # in swish-e 2.4.5
        'HTML::HiLiter'                           => 0.12,
        'Data::Page'                              => 1.0,
        'File::Spec::Functions'                   => 0,
        'Number::Format'                          => 0,
        'HTML::FillInForm'                        => 1.04,
        'Time::HiRes'                             => 0,
        'Time::Piece'                             => 0,
        'POSIX'                                   => 0,
        'Text::Context'                           => 3.5,
        'File::Spec'                              => 0,
    },
    build_requires  => {
        'Test::LongString'                        => 0,
        'Test::Class'                             => 0,
    },
    create_readme      => 1,
    pm_files => {
        'lib/CGI/Application/Search.pm'      => 'lib/CGI/Application/Search.pm',
        'templates/search_results.tmpl'      => 'lib/CGI/Application/Search/templates/search_results.tmpl',
        'templates/search_results.tt'        => 'lib/CGI/Application/Search/templates/search_results.tt',
        'templates/ajax_search_results.tmpl' => 'lib/CGI/Application/Search/templates/ajax_search_results.tmpl',
        'templates/ajax_search_results.tt'   => 'lib/CGI/Application/Search/templates/ajax_search_results.tt',
        'templates/prototype_javascript.tmpl'=> 'lib/CGI/Application/Search/templates/prototype_javascript.tmpl',
    },
)->create_build_script();


lib/CGI/Application/Search/Tutorial.pm
use strict; # this makes CPANTS happy, even though it's straight POD
=head1 NAME 

CGI::Application::Search::Tutorial - How do we use this thing?

=head1 DESCRIPTION

Need information on how to setup Swish-e for use with 
L<CGI::Application::Search>?

Need more information on using the AJAX capabilities of 
L<CGI::Application::Search>?

Then you've come to the right place

=head1 BRIEF SWISH-E TUTORIAL

You can skip this section if you're a Swish-e veteran.  Otherwise,
read on for a step-by-step guide to adding a search interface to your
site using CGI::Application::Search.

=head2 Step 1: Install Swish-e

The first thing you need to do is install Swish-e.  First, download it
from the swish-e site:

   http://swish-e.org

Then unpack it, cd into the directory, build and install:

  tar zxf swish-e-2.4.3.tar.gz
  cd swish-e-2.4.3
  ./configure
  make
  make install

You'll also need to build the Perl module, SWISH::API, which this
module uses:

  cd perl
  perl Makefile.PL
  make
  make install

=head2 Step 2: Setup a Config File

The first step to setting up a swish-e search engine is writing a
config file.  Swish-e supports a smorgasborg of configuration options
but just a few will get you started.

  # index all HTML files in /path/to/index
  IndexDir /path/to/index
  IndexOnly .html .htm
  IndexContents HTML2 .html .htm

  # C::A::Search needs a description, use the first 1,500 characters
  # of the body
  StoreDescription HTML2 <body> 1500

  # remove doc-root path so links will work on the results page
  ReplaceRules remove /path/to/index

Put the above in a file called F<swish-e.conf>.

NOTE: The above is a very simple swish-e.conf file. To bask in the
power and flexibility that is swish-e, please see the official documentation. 

=head2 Step 3: Run the Indexer

Now that you've got a basic configuration file you can index your site.  
The corresponding simple command is:

  $ swish-e -v 1 -c swish-e.conf -f /path/to/swishe-index

The last part is the place where Swish-e will write its index.  It
should be the name of a file in a directory writable by you and
readable by your CGI scripts.

Later you'll need to setup the indexer to run from cron, but for now
just run it once.

=head2 Step 4: Run a Test Search

Swish-e has a command-line interface to running searches which you can
use to confirm that your index is working.  For example, to search for
"foo":

  $ swish-e -w foo -f /path/to/swishe-index

If that works you should see some hits (assuming your site contains
"foo").

=head2 Step 5: Setup an Instance Script

Like all CGI::Application modules, CGI::Application::Search requires
an instance script.  Create a file called 'search.pl' or 'search.cgi'
in a place where your web server will execute it.  Put this in it:

  #!/usr/bin/perl -w
  use strict;
  use CGI::Application::Search;
  my $app = CGI::Application::Search->new(
    PARAMS => { SWISHE_INDEX => '/path/to/index' }
  );
  $app->run();

Now make it executable:

  $ chmod +x search.pl

=head2 Step 6: Test Your Instance Script

First, test it on the command-line:

  $ ./search.pl

That should show you the HTML for the search form with no results.
Now try it in your browser:

  http://yoursite.example.com/search.pl

If that doesn't work, check your error log.  Do not email me or the
CGI::Application mailing list until you check your error log.  Yes, I
mean you. Thanks.

=head2 Step 7: Rejoice

You've just completed the world's easiest search system setup!  Now go
setup that indexing cronjob.

=head1 AJAX USAGE

L<CGI::Application::Search> provides 2 features implemented in AJAX
(Asynchronous Javascript And XML). These are:

=over

=item Non-Refresh Search

Only the relevant portions of the page
are changed, not the entire page. This results in a faster search, especially
if the page is surrounded by other dynamic elements (navigation, side bars, etc).

=item Auto-Suggest

As the user types, they are presented with suggestions that match the letters/words
they have entered so far.

=back

Both are configurable and overrideable and can also be turned off completely.
Both make use of the B<Prototype> and B<Scriptaculous> JavaScript libraries 
(available at L<http://prototype.conio.net/> and L<http://script.aculo.us>).

Although our example AJAX templates have these libraries included (using 
C<< HTML::Prototype->define_javascript_functions() >> we recommend that you
actually download these libraries yourself and put them into your web document
tree and reference them in C<< <script> >> tags. This will allow them to be
cached by the browser instead of reparsed on each page fetch.

  <script src="/prototype.js" type="text/javascript"></script>
  <script src="/scriptaculous.js" type="text/javascript"></script>

You are encouraged to look at the sample templates included with this distribution
while you are learning how the JavaScript, CSS, C<< <forms> >> and links all work
together.

=head2 SETTING UP NON-REFRESH SEARCH

This is fairly straight forward. All form submissions and all
links are converted into calls to the C<< Ajax.Updater() >> method instead.

Form submissions will be serialized into query strings, and both links
and form submissions will be sent to the server. What ever the server
sends back will just replace the contents of the C<< <div> >> with the
id of 'search_listing'. And while the browser waits for the response,
the 'search_listing' div is replaced by a simple message.

    <div id="search_listing"></div>

    <script type="text/javascript">
    <!--
        new Ajax.Updater(
            'search_listing',
            url,
            {
                parameters: query,
                asynchronous: 1,
                onLoading: function(request) {
                    $('search_listing').innerHTML = "<strong>" + msg + " ...</strong>";
                }
            }
        );
    -->
    </script>


You can use the same template for both the inital request to view the full
search page (with search form) and to just return the Non-Refresh search
results. Simply use the 'ajax' flag to determine what to show when. See
the sample templates (F<templates/ajax_search_results.tmpl> and 
F<templates/ajax_search_results.tt>) for examples.

After you have added the appropriate C<< <div> >> and Javascript, then
simple set the B<AJAX> config parameter to true in your L<CGI::Application::Search>'s
params.

=head2 SETTING UP AUTO-SUGGEST  

By default, the AUTO-SUGGEST feature will use a simple flat file that contains
the words to suggest, in alphabetical order, one word per line. You can specify
the file (B<AUTO_SUGGEST_FILE>) and whether or not the application should cache the 
values it pulls from the file (B<AUTO_SUGGEST_CACHE>) which is useful if the file 
only contains a few thousand words and you're in a persistent environment. 

You can also specify the maximum number of suggestions that can be sent to the 
user at a given time (B<AUTO_SUGGEST_LIMIT>). This is useful to not only speed i
up the suggestions, but also keeps from overwhelming the use.

A common use-case is to suggest words that are known to exist in the Swish-e
index. After your index has been set up, it's pretty trivial to extract all
of the words that exist into a separate file perfect for use as an 
B<AUTO_SUGGEST_FILE>.

This command will create an alphabetical listing of every word in your documents
that has at least 2 alphabetical characters.

    # in bash
    swish-e -T INDEX_WORDS_META -f swishe.index \
        | grep -o -P "^\S+"  \
        | grep -P "[a-z]{2}" \
     > auto_suggest_file

Then in your template, you have to make sure that an C<< Ajax.Autocompleter() >> which
will watch the form input in question and send the requests to the search app (for the
'suggestions' run mode) and then show the user those suggestions.

    <script type="text/javascript">
    <!--
        var url = '<tmpl_var url>';
        new Ajax.Autocompleter( 
            'keywords', 
            'keywords_auto_complete', 
            url, 
            { parameters: "rm=suggestions"  }
        )
    //-->
    </script>

Next, make sure that you have the appropriate C<< <div> >> in your HTML to contain
those suggestions (most likely immediately below the form input). It would also probably
be desireable to turn the browser's built-in auto-completion off for this field.

    <input type="text" name="keywords" id="keywords" autocomplete="off" />
    <div class="auto_complete" id="keywords_auto_complete"></div>

The last step is to make sure that you have the appropriate CSS styles defined so that
your auto-suggested results can be shown in the client. The easiest to get these CSS rules
is from HTML::Prototype. Take the output from the following command, paste it into your
templates and customize to match your look and feel.

    perl -MHTML::Prototype -e 'print HTML::Prototype->auto_complete_stylesheet . "\n"';

=head1 AUTHOR

Michael Peters <mpeters@plusthree.com>

Thanks to Plus Three, LP (http://www.plusthree.com) for sponsoring my work on this module.

=head1 CONTRIBUTORS

=over

=item Sam Tregar <sam@tregar.com>

=back

=cut

# just in case someone is stupid enough to actually 'use' us :)

1;




Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.03
use ExtUtils::MakeMaker;
WriteMakefile
(
          'PL_FILES' => {},
          'INSTALLDIRS' => 'site',
          'NAME' => 'CGI::Application::Search',
          'EXE_FILES' => [],
          'VERSION_FROM' => 'lib/CGI/Application/Search.pm',
          'PM' => {
                    'lib/CGI/Application/Search.pm' => '$(INST_LIBDIR)/Search.pm',
                    'templates/prototype_javascript.tmpl' => '$(INST_LIBDIR)/Search/templates/prototype_javascript.tmpl',
                    'templates/search_results.tmpl' => '$(INST_LIBDIR)/Search/templates/search_results.tmpl',
                    'templates/ajax_search_results.tt' => '$(INST_LIBDIR)/Search/templates/ajax_search_results.tt',
                    'templates/search_results.tt' => '$(INST_LIBDIR)/Search/templates/search_results.tt',
                    'templates/ajax_search_results.tmpl' => '$(INST_LIBDIR)/Search/templates/ajax_search_results.tmpl'
                  },
          'PREREQ_PM' => {
                           'SWISH::API' => '0.01',
                           'File::Spec' => 0,
                           'Time::Piece' => 0,
                           'Number::Format' => 0,
                           'Time::HiRes' => 0,
                           'Data::Page' => '1',
                           'CGI::Application' => '3.2',
                           'CGI::Application::Plugin::HTMLPrototype' => 0,
                           'POSIX' => 0,
                           'HTML::FillInForm' => '1.04',
                           'HTML::HiLiter' => '0.12',
                           'CGI::Application::Plugin::AnyTemplate' => 0,
                           'File::Spec::Functions' => 0,
                           'Text::Context' => '3.5'
                           'Test::LongString' => 0,
                         }
        )
;
